kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+                        .joinToString()
+                    val initArgs = qualifiers
+                        .map { it.annotationType.asTypeName() }
+                        .toTypedArray()
+                    ", $initString" to initArgs
+                  }
+                }
+                val finalArgs = arrayOf(*standardArgs, *args)
+                initializer(
+                    "%1N.adapter%2L(%3L$initializerString)${if (prop.nullable) ".nullSafe()" else ""}",
+                    *finalArgs)
+              }
+              .build()
+        }
+
+    val localProperties =
+        propertyList.associate { prop ->
+          val propertySpec = PropertySpec.builder(prop.name.allocate(), prop.typeName.asNullable())
+              .mutable(true)
+              .initializer("null")
+              .build()
+          val propertySetSpec = if (prop.hasDefault && prop.nullable) {
+            PropertySpec.builder("${propertySpec.name}Set".allocate(), BOOLEAN)
+                .mutable(true)
+                .initializer("false")
+                .build()
+          } else {
+            null
+          }
+          val specs = propertySpec to propertySetSpec
+          prop to specs
+        }
+    val optionsByIndex = propertyList
+        .associateBy { it.serializedName }.entries.withIndex()
+
+    // selectName() API setup
+    val optionsCN = JsonReader.Options::class.asTypeName()
+    val optionsProperty = PropertySpec.builder(
+        "options".allocate(),
+        optionsCN,
+        PRIVATE)
+        .initializer("%T.of(${optionsByIndex.map { it.value.key }
+            .joinToString(", ") { "\"$it\"" }})",
+            optionsCN)
+        .build()
+
+    val adapter = TypeSpec.classBuilder(adapterName)
+        .superclass(jsonAdapterTypeName)
+        .apply {
+          genericTypeNames?.let {
+            addTypeVariables(genericTypeNames)
+          }
+        }
+        .apply {
+          // TODO make this configurable. Right now it just matches the source model
+          if (visibility == INTERNAL) {
+            addModifiers(KModifier.INTERNAL)
+          }
+        }
+        .primaryConstructor(FunSpec.constructorBuilder()
+            .addParameter(moshiParam)
+            .apply {
+              genericTypeNames?.let {
+                addParameter(typesParam)
+              }
+            }
+            .build())
+        .addProperty(optionsProperty)
+        .addProperties(adapterProperties.values)
+        .addFunction(FunSpec.builder("toString")
+            .addModifiers(OVERRIDE)
+            .returns(String::class)
+            .addStatement("return %S",
+                "GeneratedJsonAdapter(${originalTypeName.resolveRawType()
+                    .simpleNames()
+                    .joinToString(".")})")
+            .build())
+        .addFunction(FunSpec.builder("fromJson")
+            .addModifiers(OVERRIDE)
+            .addParameter(reader)
+            .returns(originalTypeName.asNullable())
```

Comment:
No reason to put `asNullable()` since we tell the caller to call `nullSafe()` if they want `null` support.

Reply:
960f85e